Add SystemTime::{MIN, MAX}

cvengler opened the pull request:
Accepted ACP: <https://github.com/rust-lang/libs-team/issues/692>
Tracking Issue: <https://github.com/rust-lang/rust/issues/149067>

---

This merge request introduces two new constants to `SystemTime`: `MIN` and `MAX`, whose values represent the maximum values for the respective data type, depending upon the platform.

Technically, this value is already obtainable during runtime with the following algorithm:
Use `SystemTime::UNIX_EPOCH` and call `checked_add` (or `checked_sub`) repeatedly with `Duration::new(0, 1)` on it, until it returns None.
Mathematically speaking, this algorithm will terminate after a finite amount of steps, yet it is impractical to run it, as it takes practically forever.

Besides, this commit also adds a unit test to verify those values represent the respective minimum and maximum, by letting a `checked_add` and `checked_sub` on it fail.

In the future, the hope of the authors lies within the creation of a `SystemTime::saturating_add` and `SystemTime::saturating_sub`, similar to the functions already present in `std::time::Duration`.
However, for those, these constants are crucially required, thereby this should be seen as the initial step towards this direction.
With this change, implementing these functions oneself outside the standard library becomes feasible in a portable manner for the first time.

This feature (and a related saturating version of `checked_{add, sub}` has been requested multiple times over the course of the past few years, most notably:
* rust-lang/rust#100141
* rust-lang/rust#133525
* rust-lang/rust#105762
* rust-lang/rust#71224
* rust-lang/rust#45448
* rust-lang/rust#52555

cvengler committed:
time: Implement SystemTime::{MIN, MAX}

This commit introduces two new constants to SystemTime: `MIN` and `MAX`,
whose value represent the maximum values for the respective data type,
depending upon the platform.

Technically, this value is already obtainable during runtime with the
following algorithm: Use `SystemTime::UNIX_EPOCH` and call `checked_add`
(or `checked_sub`) repeatedly with `Duration::new(0, 1)` on it, until it
returns None.  Mathematically speaking, this algorithm will terminate
after a finite amount of steps, yet it is impractical to run it, as it
takes practically forever.

Besides, this commit also adds a unit test.  Concrete implementation
depending upon the platform is done in later commits.

In the future, the hope of the authors lies within the creation of a
`SystemTime::saturating_add` and `SystemTime::saturating_sub`, similar
to the functions already present in `std::time::Duration`.  However, for
those, these constants are crucially required, thereby this should be
seen as the initial step towards this direction.

Below are platform specifc notes:

# Hermit

The HermitOS implementation is more or less identitcal to the Unix one.

# sgx

The implementation uses a `Duration` to store the Unix time, thereby
implying `Duration::ZERO` and `Duration::MAX` as the limits.

# solid

The implementation uses a `time_t` to store the system time within a
single value (i.e. no dual secs/nanosecs handling), thereby implying its
`::MIN` and `::MAX` values as the respective boundaries.

# UEFI

UEFI has a weird way to store times, i.e. a very complicated struct.
The standard proclaims "1900-01-01T00:00:00+0000" to be the lowest
possible value and `MAX_UEFI_TIME` is already present for the upper
limit.

# Windows

Windows is weird.  The Win32 documentation makes no statement on a
maximum value here.  Next to this, there are two conflicting types:
`SYSTEMTIME` and `FILETIME`.  Rust's Standard Library uses `FILETIME`,
whose limit will (probably) be `i64::MAX` packed into two integers.
However, `SYSTEMTIME` has a lower-limit.

# xous

It is similar to sgx in the sense of using a `Duration`.

# unsupported

Unsupported platforms store a `SystemTime` in a `Duration`, just like
sgx, thereby implying `Duration::ZERO` and `Duration::MAX` as the
respective limits.

Kivooeo commented:
I'm pretty sure, though I could be wrong that this will need an [ACP](https://github.com/rust-lang/libs-team/issues/new?template=api-change-proposal.md) first before any work

cvengler commented:
Marking this as ready for review, given that CI works now. 🎉

cvengler commented:
ACP opened: https://github.com/rust-lang/libs-team/issues/692

ChrisDenton commented on library/std/src/sys/pal/hermit/time.rs:
```diff
@@ -15,6 +15,12 @@ struct Timespec {
 }
 
 impl Timespec {
+    #[unstable(feature = "time_systemtime_limits", issue = "none")]
```
You shouldn't need to add the `#[unstable]` for internal types in `sys/`. Only to the public facing type.

cvengler replied:
fixed in 5f1a63f015cb5e916c4b1e2fde30921d5fa96c03

ChrisDenton reviewed (commented):
The ACP has been accepted! Once the tracking issue is created please update the issue number in the `unstable` attributes.

cvengler commented:
> The ACP has been accepted! Once the tracking issue is created please update the issue number in the `unstable` attributes.

Thank you for the super fast review!
Already created the issue but I cannot push the commit adding due to GitHub SSH being down for me (probably due to the CloudFlare incident). Will try again in some time.

joboet commented on library/std/src/sys/pal/unsupported/time.rs:
```diff
@@ -27,6 +27,10 @@ impl Instant {
 impl SystemTime {
+    pub const MAX: SystemTime = Duration::ZERO;
+
+    pub const MIN: SystemTime = Duration::MAX;
```
I believe these two are mixed up?

cvengler replied:
fixed in 348d68443cfaa067d5235a8d7606214cc56bb999

cvengler commented on library/std/src/sys/pal/windows/time.rs:
```diff
@@ -64,6 +64,22 @@ impl Instant {
+        t: c::FILETIME {
+            dwLowDateTime: (i64::MIN & 0xFFFFFFFF) as u32,
+            dwHighDateTime: (i64::MIN >> 32) as u32,
+        },
```
Nevermind, it seems to work! 🙂

cvengler commented:
@rustbot ready

madsmtm commented on library/std/src/time.rs:
```diff
@@ -511,6 +511,46 @@ impl SystemTime {
+    ///
+    /// This value differs a lot between platforms, but it is always the case
+    /// that any positive subtraction from [`SystemTime::MIN`] will fail.
+    ///
```
Q: Should this link to `SystemTime::UNIX_EPOCH`, or somehow explain the difference between the two?

madsmtm commented on library/std/src/time.rs:
```diff
@@ -511,6 +511,46 @@ impl SystemTime {
+    /// This value differs a lot between platforms, but it is always the case
+    /// that any positive subtraction from [`SystemTime::MIN`] will fail.
+    ///
+    /// # Examples
```
Remark: I think the motivating example you gave in https://github.com/rust-lang/libs-team/issues/692 is useful in describing why you'd actually need this. Could we add something like that too?

madsmtm reviewed (changes requested):
Thanks for the ping. I think this makes sense to have, but I'm a bit worried about the exact bounds that we expose, especially given that the `Timespec` that the `sys/pal/unix/time.rs` module uses has a 64-bit wide `tv_sec`, while `libc::timespec`'s may only be 32-bit. See also the first comment.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148825/3d461af2a23456a2676aadb13b4253c87bdfe28d..348d68443cfaa067d5235a8d7606214cc56bb999)*

ijackson commented on library/std/tests/time.rs:
```diff
@@ -237,9 +238,27 @@ fn system_time_duration_since_max_range_on_unix() {
+    assert!(SystemTime::MAX.checked_sub(Duration::new(0, 1)).is_some());
+    assert!(SystemTime::MIN.checked_add(Duration::new(0, 1)).is_some());
+    assert!(SystemTime::MIN.checked_sub(Duration::new(0, 1)).is_none());
+}
```
> I don't actually think this will work on platforms where `time_t` is 32-bit.

I'm not sure what you mean by "this" (or "work")  :-).

According to my reading of `unix/time.rs`, on systems where `time_t` is 32-bit, one can still call `Duration::from_secs` on a 64-bit value and `checked_add` that to `UNIX_EPOCH` and get a `SystemTime` containing the 64-bit value.

One can also observe that `(UNIX_EPOCH + Duration::from_secs(i64::MAX)).checked_add(Duration::from_nanos(1)` is `None`.

Functions that try to interface with the OS will have to do "something" with the 64-bit value.  I haven't investigated what that is.

All of this is true today.  This MR doesn't touch it.  If this is a bug, it's a pre-existing bug.

I think we don't want `SystemTime::MAX` to be based on 32-bit arithmetic., on any platform.  Certainly it would be bizarre to have `SystemTime::MAX` be 32-bit but to be able to create values greater than `MAX` by addition!

(Eited to fix a u64 that ought to have been i64, and fix parens.)

madsmtm replied:
> All of this is true today. This MR doesn't touch it. If this is a bug, it's a pre-existing bug.

True, but these interactions and limits probably weren't thought of that much before, which is why it makes sense to consider them now.

> I think we don't want `SystemTime::MAX` to be based on 32-bit arithmetic., on any platform.

I disagree. I think it makes sense to have `SystemTime::MIN`/`SystemTime::MAX` be "the boundaries in the system APIs we use", not just "some arbitrary[^1] value used in the current Rust implementation".

For the platforms that have a 32-bit signed `tv_sec`, `SystemTime::now()` will on January 19th 2038 start returning weird values one way or another. We can't really do anything about that.

Once that day comes though, I think it makes much more sense that they start returning `SystemTime::MIN + delta` instead of `SystemTime::UNIX_EPOCH - i32::MIN + delta`.

[^1]: For comparison, consider: Why are we even using `i64` + `u32` nanos, instead of just counting the number of nanoseconds in a `i64`, that could presumably also be a valid implementation, and that'd change the bounds.

madsmtm commented on library/std/src/sys/pal/unix/time.rs:
```diff
@@ -61,6 +65,13 @@ impl fmt::Debug for SystemTime {
 impl Timespec {
+    const MAX: Timespec = unsafe { Self::new_unchecked(i64::MAX, 1_000_000_000 - 1) };
+
+    // As described below, on AppleOS, dates before epoch are represented differently.
```
At least use `Apple OS` (with the space) as below then, `AppleOS` sounds too much like `macOS` / `tvOS` etc.

cvengler replied:
fixed in 2a2c12036bd

cvengler commented:
> Given that this is nightly only and libs-api have accepted this in general, I'm ok with merging the current implementation and letting the discussion continue on the tracking issue. If there are more issues that need to be brought to the attention of libs-api then the tracking issue can be nominated for a team discussion.
> 
> Could you squish your commits?

done!

matthiaskrgr commented:
@bors r-
https://github.com/rust-lang/rust/actions/runs/20146572689/job/57828597532#step:29:9732

https://github.com/rust-lang/rust/pull/149890#issuecomment-3643787524

JonathanBrouwer commented:
@bors try jobs=dist-various-2

ChrisDenton commented:
@bors try jobs=dist-various-2

cvengler committed:
time: Fix Windows' `SystemTime::checked_sub`

The Windows implementation of `SystemTime::checked_sub` contains a bug,
namely that it does not return `None` on values below 1601.

This bug stems from the fact that internally, the time gets converted to
an i64, with zero representing the anchor in 1601.  Of course,
performing checked subtraction on a signed integer generally works fine.
However, the resulting value delivers undefined behavior on Windows
systems.

To mitigate this issue, we try to convert the resulting i64 to an u64
because a negative value should obviously fail there.

cvengler committed:
time: Test and document time precision edge-case

There is a slight edge case when adding and subtracting a `Duration`
from a `SystemTime`, namely when the duration itself is finer/smaller
than the time precision on the operating systems.

On most (if not all non-Windows) operating systems, the precision of
`Duration` aligns with the `SystemTime`, both being one nanosecond.

However, on Windows, this time precision is 100ns, meaning that adding
or subtracting a `Duration` whose value is `< Duration::new(0, 100)`
will result in that method behaving like an addition/subtracting of
`Duration::ZERO`, due to the `Duration` getting rounded-down to the zero
value.

cvengler commented:
> > Use SystemTime::UNIX_EPOCH and call checked_add (or checked_sub) repeatedly with Duration::new(0, 1) on it, until it returns None.
> > Mathematically speaking, this algorithm will terminate after a finite amount of steps, yet it is impractical to run it, as it takes practically forever.
> 
> I do think it's a neat addition, but I do want to note that there are efficient ways to determine the min & max at runtime.
> 
> [With binary search](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=21b23046027e700a8dca1a809fa22489), it only takes 96 calls to `checked_sub` to compute the minimum on the playground:
> 
> ```rust
> fn min() -> SystemTime {
>     let mut t = SystemTime::UNIX_EPOCH;
>     let mut d = Duration::MAX;
>     
>     while d > Duration::ZERO {
>         while let Some(candidate) = t.checked_sub(d) {
>             t = candidate;
>         }
>         
>         d /= 2;
>     }
>     
>     t
> }
> ```

Also check out `saturating-time` on crates.io
